File: First-Camp/leetcode/98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def preOrder(root, minVal, maxVal):
            if not root: return True
            if root.val >= maxVal or root.val <= minVal: return False

            left = preOrder(root.left, minVal, root.val)
            right = preOrder(root.right, root.val, maxVal)
            return left and right
        return preOrder(root, float('-inf'), float('inf'))

File: Second-Camp/leetcode/98. Validate Binary Search Tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool preorder(TreeNode* root, long minval, long maxval) {
        if (!root) return true;
        if (root->val >= maxval || root->val <= minval) return false;
        bool left = preorder(root->left, minval, root->val);
        bool right = preorder(root->right, root->val, maxval);
        return left&&right;
    }
    bool isValidBST(TreeNode* root) {
        return preorder(root, LONG_MIN, LONG_MAX);
    }
};
